tests/memory/enc/fragment.c

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+#include <openenclave/internal/globals.h>
+#include <openenclave/internal/tests.h>
+
+#include <malloc.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "memory_t.h"
+
+#define MALLOC_SIZE_SMALL 1024
+#define EFFICIENCY_TEST_TIMES 5000000
+
+#define MALLINFO_FIELD_TYPE size_t
+struct mallinfo
+{
+    MALLINFO_FIELD_TYPE arena;    /* non-mmapped space allocated from system */
+    MALLINFO_FIELD_TYPE ordblks;  /* number of free chunks */
+    MALLINFO_FIELD_TYPE smblks;   /* always 0 */
+    MALLINFO_FIELD_TYPE hblks;    /* always 0 */
+    MALLINFO_FIELD_TYPE hblkhd;   /* space in mmapped regions */
+    MALLINFO_FIELD_TYPE usmblks;  /* maximum total allocated space */
+    MALLINFO_FIELD_TYPE fsmblks;  /* always 0 */
+    MALLINFO_FIELD_TYPE uordblks; /* total allocated space */
+    MALLINFO_FIELD_TYPE fordblks; /* total free space */
+    MALLINFO_FIELD_TYPE keepcost; /* releasable (via malloc_trim) space */
+};
+
+struct mallinfo dlmallinfo(void);
+
+static size_t _get_heap_size()
+{
+    // call malloc to update the globlal data.
+    free(malloc(1));
+    return dlmallinfo().keepcost;
+}
+
+static int _malloc_free_fixed_size(size_t size, int times)
+{
+    int i = 0;
+    int* buffer = NULL;
+
+    for (i = 0; i < times; i++)
+    {
+        buffer = (int*)malloc(size);
+        if (NULL == buffer)
+        {
+            return -1;
+        }
+        free(buffer);
+    }
+
+    oe_host_printf("_malloc_free_fixed_size malloc times = %d.\n", i);
+    return 0;
+}
+
+static inline size_t _randx(size_t x)
+{
+    /* Note that rand() % N is biased if RAND_MAX + 1 isn't divisible
+     * by N. But, slight probability bias doesn't really matter in these
+     * tests. */
+    return (size_t)rand() % x;
+}
+
+static int _malloc_free_random_size(int times)
+{
+    int i = 0;
+    size_t size = 0;
+    void* buffer = NULL;
+
+    for (i = 0; i < times; i++)
+    {
+        size = _randx(_get_heap_size());
```

Comment:
Having a test with random behaviour concerns me since regressions may not be reproducible. I understand you want to test many different values. Could we seed the RNG at the beginning of the test so that the test's behavior is deterministic?

Reply:
The purpose of this case is to test random behavior. because the user cannot determine the memory usage behavior.

Reply:
I'm looking for deterministic randomness. The goal of the test to to try a bunch a different unexpected values. It doesn't really matter if those values change on each new run of the test.

If you seed the rng at the beginning of the test with a constant, you'll still get random value, but those random values won't change on each new run. This makes the test reproducible and regressions will also be reproducible.

Reply:
Thanks jhand2 for your suggestion. this method is great for reproducible.  I concerned whether using fixed sequences may cover up the problem. maybe this sequences is fine ,but another is not.

Reply:
In my opinion having a reproducible test is better than greater coverage because having tests with large coverage that can't be reproduced makes it hard to fix the problems detected.

What do you think?

Reply:
@yanxue-intel another option would be to generate a random seed at the beginning of the test and print that seed. Then the test could accept a command-line parameter to use a passed seed instead of randomly generating one.

That way if we see a failure with a given seed, we can pass in that seed to repro the failure. This gives us both benefits.

Reply:
> can't



> In my opinion having a reproducible test is better than greater coverage because having tests with large coverage that can't be reproduced makes it hard to fix the problems detected.
> 
> What do you think?

agree with you.